Linux Build - Py3.11

Hello Jinmo;

In azure pipelines I could not found Linux binaries, am I right?

I wonder how can I build plugin for linux! Thank you.

Platform: Arch GNU/Linux x86_64

First I have cloned the repo with recursive:

```
git clone https://github.com/Jinmo/ifred.git --recursive
```

Then cd into repo and make a build dir:

```
cd ifred
mkdir build
cd build
```

Then I have first run cmake. With these cmake parameters:

```
#!/bin/sh

export IDA_SDK="/opt/idapro82/sdk"
export IDA_INSTALL_DIR="/opt/idapro82"

cmake ../ \
    -DIDA_SDK="$IDA_SDK" \
    -DIDA_BINARY_64=True \
    -DIDA_EA_64=False \
    -GNinja \
    -DPYBIND11_PYTHON_VERSION=3 \
    -DIDA_INSTALL_DIR="$IDA_INSTALL_DIR"
```

I got got this output:

```
- The C compiler identification is GNU 13.2.1
-- The CXX compiler identification is GNU 13.2.1
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /usr/bin/cc - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /usr/bin/c++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
CMake Deprecation Warning at palette/pybind11/CMakeLists.txt:8 (cmake_minimum_required):
  Compatibility with CMake < 3.5 will be removed from a future version of
  CMake.

  Update the VERSION argument <min> value or use a ...<max> suffix to tell
  CMake that the project does not need compatibility with older versions.


-- pybind11 v2.10.2
CMake Warning (dev) at palette/pybind11/tools/FindPythonLibsNew.cmake:98 (find_package):
  Policy CMP0148 is not set: The FindPythonInterp and FindPythonLibs modules
  are removed.  Run "cmake --help-policy CMP0148" for policy details.  Use
  the cmake_policy command to set the policy and suppress this warning.

Call Stack (most recent call first):
  palette/pybind11/tools/pybind11Tools.cmake:50 (find_package)
  palette/pybind11/tools/pybind11Common.cmake:180 (include)
  palette/pybind11/CMakeLists.txt:208 (include)
This warning is for project developers.  Use -Wno-dev to suppress it.

-- Found PythonInterp: /usr/bin/python3 (found suitable version "3.11.7", minimum required is "3")
-- Found PythonLibs: /usr/lib/libpython3.11.so
-- Performing Test HAS_FLTO
-- Performing Test HAS_FLTO - Success
CMake Deprecation Warning at palette/CMakeLists.txt:1 (cmake_minimum_required):
  Compatibility with CMake < 3.5 will be removed from a future version of
  CMake.

  Update the VERSION argument <min> value or use a ...<max> suffix to tell
  CMake that the project does not need compatibility with older versions.


CMake Deprecation Warning at palette/ida-cmake/cmake/IDA.cmake:25 (cmake_minimum_required):
  Compatibility with CMake < 3.5 will be removed from a future version of
  CMake.

  Update the VERSION argument <min> value or use a ...<max> suffix to tell
  CMake that the project does not need compatibility with older versions.
Call Stack (most recent call first):
  palette/CMakeLists.txt:4 (include)


-- Performing Test COMPILER_HAS_HIDDEN_VISIBILITY
-- Performing Test COMPILER_HAS_HIDDEN_VISIBILITY - Success
-- Performing Test COMPILER_HAS_HIDDEN_INLINE_VISIBILITY
-- Performing Test COMPILER_HAS_HIDDEN_INLINE_VISIBILITY - Success
-- Performing Test COMPILER_HAS_DEPRECATED_ATTR
-- Performing Test COMPILER_HAS_DEPRECATED_ATTR - Success
CMake Deprecation Warning at palette/ida-cmake/cmake/IDA.cmake:25 (cmake_minimum_required):
  Compatibility with CMake < 3.5 will be removed from a future version of
  CMake.

  Update the VERSION argument <min> value or use a ...<max> suffix to tell
  CMake that the project does not need compatibility with older versions.
Call Stack (most recent call first):
  palette/src/ida/CMakeLists.txt:1 (include)


-- Configuring done (8.0s)
-- Generating done (0.1s)
-- Build files have been written to: /home/mypathto/jinmo/ifred/build
```

Then finally I ran `ninja`,  and got this error:


```
ninja
[17/18] Building CXX object palette/src/ida/CMakeFiles/ida_palette.dir/plugin.cpp.o
FAILED: palette/src/ida/CMakeFiles/ida_palette.dir/plugin.cpp.o
/usr/bin/c++ -DNO_OBSOLETE_FUNCS -DPYTHON_SUPPORT=1 -DQT_CORE_LIB -DQT_GUI_LIB -DQT_NAMESPACE=QT -DQT_NO_DEBUG -DQT_WIDGETS_LIB -D__IDP__ -D__LINUX__ -D__X64__ -Dida_palette_EXPORTS -I/home/reverser/DEV/Github/jinmo/ifred/build/palette/src/ida -I/home/reverser/DEV/Github/jinmo/ifred/palette/src/ida -I/home/reverser/DEV/Github/jinmo/ifred/build/palette/src/ida/ida_palette_autogen/include -I/opt/idapro82/sdk/include -I/opt/idapro82/plugins/hexrays_sdk/include -I/home/reverser/DEV/Github/jinmo/ifred/palette/include -I/home/reverser/DEV/Github/jinmo/ifred/build/palette -isystem /usr/include/qt -isystem /usr/include/qt/QtCore -isystem /usr/lib/qt/mkspecs/linux-g++ -isystem /usr/include/qt/QtGui -isystem /usr/include/qt/QtWidgets -isystem /home/reverser/DEV/Github/jinmo/ifred/palette/pybind11/include -isystem /usr/include/python3.11 -std=gnu++17 -fPIC -fPIC -MD -MT palette/src/ida/CMakeFiles/ida_palette.dir/plugin.cpp.o -MF palette/src/ida/CMakeFiles/ida_palette.dir/plugin.cpp.o.d -o palette/src/ida/CMakeFiles/ida_palette.dir/plugin.cpp.o -c /home/reverser/DEV/Github/jinmo/ifred/palette/src/ida/plugin.cpp
In file included from /home/reverser/DEV/Github/jinmo/ifred/palette/src/ida/plugin.h:22,
                 from /home/reverser/DEV/Github/jinmo/ifred/palette/src/ida/plugin.cpp:1:
/opt/idapro82/sdk/include/ida.hpp:1424:3: error: ‘uintptr_t’ does not name a type
 1424 |   uintptr_t offset;             ///< field position:
      |   ^~~~~~~~~
/opt/idapro82/sdk/include/ida.hpp:13:1: note: ‘uintptr_t’ is defined in header ‘<cstdint>’; did you forget to ‘#include <cstdint>’?
   12 | #include <range.hpp>
  +++ |+#include <cstdint>
   13 |
/opt/idapro82/sdk/include/ida.hpp: In member function ‘size_t idbattr_info_t::ridx() const’:
/opt/idapro82/sdk/include/ida.hpp:1490:71: error: ‘offset’ was not declared in this scope; did you mean ‘off_t’?
 1490 |   size_t ridx() const          { QASSERT(1932, is_node_var()); return offset; }
      |                                                                       ^~~~~~
      |                                                                       off_t
/opt/idapro82/sdk/include/ida.hpp: In member function ‘const char* idbattr_info_t::hashname() const’:
/opt/idapro82/sdk/include/ida.hpp:1491:81: error: ‘offset’ was not declared in this scope; did you mean ‘off_t’?
 1491 |   const char *hashname() const { QASSERT(2040, is_hash()); return (const char *)offset; }
      |                                                                                 ^~~~~~
      |                                                                                 off_t
ninja: build stopped: subcommand failed.
```

--- a/palette/src/ida/plugin.h
+++ b/palette/src/ida/plugin.h
@@ -19,6 +19,7 @@
 
 #define __DEFINE_PLUGIN_RETURN_CODES__
 
+#include <cstdint>
 #include <pro.h>
 #include <ida.hpp>
 #include <idp.hpp>

